File: src/converter/plugins/convertTypeParameterDeclaration.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {ifPresent} from "../render.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {VarianceModifierService, varianceModifierServiceKey} from "./VarianceModifierPlugin.js";

export const convertTypeParameterDeclaration = createSimplePlugin((node, context, render) => {
    if (!ts.isTypeParameterDeclaration(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    const varianceModifierService = context.lookupService<VarianceModifierService>(varianceModifierServiceKey)

    const name = render(node.name)

    const varianceModifier = varianceModifierService?.resolveVarianceModifier(node, context)
    const constraintType = node.constraint && render(node.constraint)
    const defaultType = node.default && render(node.default)

    return `${ifPresent(varianceModifier, it => `${it} `)}${name}${ifPresent(constraintType, it => ` : ${it}`)}${ifPresent(defaultType, it => ` /* default is ${it} */`)}`
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertTypeParameterDeclaration.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.ifPresent
import typescript.isTypeParameterDeclaration

val convertTypeParameterDeclaration = createPlugin plugin@{ node, context, render ->
    if (!isTypeParameterDeclaration(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    val varianceModifierService = context.lookupService<VarianceModifierService>(varianceModifierServiceKey)

    val name = render(node.name)

    val varianceModifier = varianceModifierService?.resolveVarianceModifier(node, context)
    val constraintType = node.constraint?.let { render(it) }
    val defaultType = node.default?.let { render(it) }

    "${ifPresent(varianceModifier) { "$it "}}${name}${ifPresent(constraintType) { " : $it"}}${ifPresent(defaultType) { " /* default is $it */"}}"
}
